src/public/js/util.js

```diff
@@ -112,39 +124,110 @@ var UTIL = (function() {
     if (t < 0) {
       t = t + 255;
     }
-    return exp3[t];
+    return exp3.get(t);
   }
 
-  function fieldMultiply(mult1, mult2) {
-    if (mult1 == 0 || mult2 == 0) {
-      return 0;
+  // Trims leading zeros off of an array
+  // arr should be an array or typed array of integers
+  function trimLeadingZeros(arr) {
+    for (var i = 0; i < arr.length; i++) {
+      if (arr[i] != 0) {
+        return arr.slice(i);
+      }
     }
-    var t = log3[mult1] + log3[mult2];
-    if (t > 255) {
-      t = t - 255;
+    return new arr.constructor([0]);
+  }
+
+  // Returns sum of two polynomials in GF(256).
+  // Both addend1 and addend2 are big-endian arrays of integers which
+  // represent polynomials (so, [1,2,3] is x^2 + 2*x + 3)
+  function polynomialAdd(addend1, addend2) {
+    // Copy the longer array as the result & name the shorter array the addend
+    var result;
+    var addend;
+    if (addend1.length > addend2.length) {
+      result = addend1.slice(0);
+      addend = addend2;
+    } else {
+      result = addend2.slice(0);
+      addend = addend1;
+    }
+
+    for (var i = 1; i <= addend.length; i++) {
+      result[result.length - i] ^= addend[addend.length - i]; // Galois extension field addition is xor with base field 2
     }
-    return exp3[t];
+
+    return trimLeadingZeros(result);
   }
 
-  function shift(msg, rot) {
-    return msg.split('').reduce(function(accumulator, letter, index) {
-      var shiftedCharCode = letter.charCodeAt() + rot;
-      var shiftedLetter = String.fromCharCode(shiftedCharCode);
-      return accumulator + shiftedLetter;
-    }, '');
+  // Returns product of two polynomials in GF(256).
+  // Both mult1 and mult2 are big-endian arrays of integers which
+  // represent polynomials (so, [1,2,3] is x^2 + 2*x + 3)
+  function polynomialMult(mult1, mult2) {
+    mult1 = trimLeadingZeros(mult1);
+    mult2 = trimLeadingZeros(mult2);
+
+    // Handle the zero case
+    if (mult1.length == 0 || mult2.length == 0) {
+      return new Uint8Array([0]);
+    }
+
+    // Each polynomial mult has degree mult.length - 1;
+    // Multiplying a polynomial of degree (a-1) by a polynomial of degree (b-1)
+    // gives a polynomial of degree (a-1)+(b-1) = a + b - 2
+    // ...which corresponds to a polynomial array of length a + b - 1
+    var resultLength = mult1.length + mult2.length - 1;
+    var result = new Uint8Array(resultLength);
+
+    // Multiply the elements of the polynomials pairwise
+    var foilCoefficient;
+    var foilPower;
+    var foilIndex;
+    for (var i = 0; i < mult1.length; i++) {
+      for (var j = 0; j < mult2.length; j++) {
+        foilCoefficient = fieldMult(mult1[i], mult2[j]);
+        foilPower = (mult1.length - 1 - i) + (mult2.length - 1 - j);
+        foilIndex = resultLength - 1 - foilPower;
+        result[foilIndex] ^= foilCoefficient; // Galois extension field addition is xor with base field 2
+      }
+    }
+
+    return trimLeadingZeros(result);
```

Comment:
Can there be a leading 0 here after trimming the previous leading 0s? I don't think so but I am not certain. If not might as well just return result to avoid needless copy.